### PR TITLE
#525 Add runs to landing page

### PR DESCRIPTION
**Pull request by lokm01**
Signed-off-by: Jan Scherbaum <kmoj02@gmail.com>

### CODE REVIEW

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala [hunk 1]
```diff
@@ -53,6 +55,51 @@ class RunMongoRepository @Autowired()(mongoDb: MongoDatabase)
...
+
+  def getTodaysStdSuccessRuns(): Future[Int] = {
+    getTodayRunsCount(List(Filters.eq("runStatus.status", "stageSucceeded")))
+  }
+
+  def getTodaysRunningRuns(): Future[Int] = {
+    getTodayRunsCount(List(Filters.eq("runStatus.status", "running")))
+  }
+
+  def getTodaysSuccessWithErrors(): Future[Int] = {
```
Comment (DzMakatun) [inline menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: Having this another status is a great idea, as it will draw more attention to the errors. Maybe it would be better to implement it in std/cnfrm jobs, so that Menas UI and the database have the same set of statuses? Otherwise, it may be misleading when tracing the errors. Alternatively, "Completed jobs with data errors", could be displayed as a separate metrics, not to be confused with job execution status... What are your thoughts on this?

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala [hunk 2]
```diff
@@ -53,6 +55,51 @@ class RunMongoRepository @Autowired()(mongoDb: MongoDatabase)
     excludeId()
   ))
 
+  private def getTodaysFilter() = {
+    val date = LocalDate.now().format(DateTimeFormatter.ofPattern("dd-MM-yyyy"))
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: This gives you local date where Menas is deployed right? Is the system-wide timezone set before this or used somehow?

File: menas/src/test/scala/za/co/absa/enceladus/menas/integration/repositories/RunRepositoryIntegrationSuite.scala [hunk 3]
```diff
@@ -44,6 +46,8 @@ class RunRepositoryIntegrationSuite extends BaseRepositoryTest {
     runFixture.dropCollection()
   }
 
+  private val today = LocalDate.now().format(DateTimeFormatter.ofPattern("dd-MM-yyyy"))
```
Comment (Zejnilovic) [inline menas/src/test/scala/za/co/absa/enceladus/menas/integration/repositories/RunRepositoryIntegrationSuite.scala]: This does not test the edge cases of what happens close to midnight

File: menas/ui/service/GenericService.js [hunk 4]
```diff
@@ -33,9 +33,40 @@ var GenericService = new function () {
     })
   };
 
+  this.runStatusFormatter = function(sStatus) {
+    if(sStatus === "failed") return "Failed";
+    else if(sStatus === "successful") return "Successfull";
```
Comment (Zejnilovic) [inline menas/ui/service/GenericService.js]:
```suggestion
    else if(sStatus === "successful") return "Successful";
```

File: menas/ui/service/GenericService.js [hunk 5]
```diff
@@ -33,9 +33,40 @@ var GenericService = new function () {
     })
   };
 
+  this.runStatusFormatter = function(sStatus) {
```
Comment (Zejnilovic) [inline menas/ui/service/GenericService.js]: JavaScript does have a switch statement, that could be better here and is supposedly faster

File: menas/ui/components/home/landingPage.view.xml [hunk 6]
```diff
@@ -11,40 +12,64 @@
...
+                                </TileContent>
+                            </tileContent>
+                        </GenericTile>
+                        <GenericTile header="Runs" class="sapUiResponsiveMargin" press="masterNavigate">
+                            <customData>
+                                <core:CustomData key="target" value="runs"></core:CustomData>
+                            </customData>
+                            <tileContent>
+                                <TileContent>
+                                    <NumericContent icon="sap-icon://physical-activity" value="{/landingPageInfo/totalNumberRuns}"></NumericContent>
```
[the remark this answers is not in the document]
Comment (lokm01) [reply menas/ui/components/home/landingPage.view.xml]: Good point

[the remark this answers is not in the document]
Comment (lokm01) [reply menas/ui/components/home/landingPage.view.xml]: Short number format will do here.

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala (hunk 1 above)
Comment (lokm01) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: @aaishaosman @yruslan 

So I guess we'd like to have as much metadata as possible. I believe in future we may need to show stuff like break down of errors etc.

**Comment by lokm01** [comment]
> 1. Could we maybe sort the graph differently? This seems hard to read
> 2. When the session expires and I re-login, I get an error in the console and buttons don't work properly until refresh.
>    image

1) Feel free to suggest.
2) Seems like we're gonna have to be a bit more deffensive with ui5 wrapper for charts.js

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala (hunk 2 above)
Comment (lokm01) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: Uses OS settings.

File: menas/src/test/scala/za/co/absa/enceladus/menas/integration/repositories/RunRepositoryIntegrationSuite.scala (hunk 3 above)
Comment (lokm01) [reply menas/src/test/scala/za/co/absa/enceladus/menas/integration/repositories/RunRepositoryIntegrationSuite.scala]: Feel free to contribute or file an issue

File: menas/ui/service/GenericService.js (hunk 4 above)
Comment (lokm01) [reply menas/ui/service/GenericService.js]: Sigh at some stage we should probably start using i18n

File: menas/ui/service/GenericService.js (hunk 5 above)
Comment (lokm01) [reply menas/ui/service/GenericService.js]: Which also bloats the code significantly (as much as I'd prefer to use a construct closer to pattern matching).

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala (hunk 2 above)
Comment (Zejnilovic) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: But that is bad, isn't it? OS settings are who knows what and we have everything in Run model in UTC.

**Comment by Zejnilovic** [comment]
> 1. Feel free to suggest.
> 2. Seems like we're gonna have to be a bit more deffensive with ui5 wrapper for charts.js

1. More like if @DzMakatun could tell you his style/logic so it corresponds
2. I am still getting these
image

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala (hunk 2 above)
Comment (Zejnilovic) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: Created #541

File: menas/src/test/scala/za/co/absa/enceladus/menas/integration/repositories/RunRepositoryIntegrationSuite.scala (hunk 3 above)
Comment (Zejnilovic) [reply menas/src/test/scala/za/co/absa/enceladus/menas/integration/repositories/RunRepositoryIntegrationSuite.scala]: Created #541 then this can be updated. Currently, it would just fail

File: menas/ui/service/GenericService.js (hunk 5 above)
Comment (Zejnilovic) [reply menas/ui/service/GenericService.js]: I think there is less tbh. I know it is 2 lines longer, but in general less useless text.
```javascript
switch(sStatus) {
    case "failed":                  return "Failed";
    case "successful":              return "Successfull";
    case "successfulWithErrors":    return "Sucessful with errors";
    case "running":                 return "Running";
    case "stdSuccessful":           return "Standardization successful";
    default:                        return sStatus;
}
```

And you don't have to tab it if you don't want to.

Comment (Zejnilovic) [reply menas/ui/service/GenericService.js]: Or check the approach in this [totaly unsuspicious link](https://ultimatecourses.com/blog/deprecating-the-switch-statement-for-object-literals#Object_Literal_lookups)

**Comment by lokm01** [comment]
The latest changes should also fix #491

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala [hunk 7]
```diff
@@ -53,6 +55,51 @@ class RunMongoRepository @Autowired()(mongoDb: MongoDatabase)
...
+    getTodayRunsCount(List(Filters.eq("runStatus.status", "running")))
+  }
+
+  def getTodaysSuccessWithErrors(): Future[Int] = {
+    getTodayRunsCount(List(
+        Filters.eq("runStatus.status", "allSucceeded"),
+        or(
+        and(Filters.exists("controlMeasure.metadata.additionalInfo.std_errors_count"),
+            Filters.notEqual("controlMeasure.metadata.additionalInfo.std_errors_count", "0")),
+        and(Filters.exists("controlMeasure.metadata.additionalInfo.std_errors_count"),
```
Comment (yruslan) [inline menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]:
```suggestion
        and(Filters.exists("controlMeasure.metadata.additionalInfo.conform_errors_count"),
```

Comment (Zejnilovic) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: this will then require a change of the names of the columns in graphs.

Comment (lokm01) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: > this will then require a change of the names of the columns in graphs.

Why?

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala (hunk 1 above)
Comment (yruslan) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: Yeah, breakdown of errors is something I'd expect in the future. I'm wondering what monitoring guys are going to say they want.

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala (hunk 7 above)
Comment (Zejnilovic) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/RunMongoRepository.scala]: We are speaking about standardization there. Aren't we? I don't have it built currently. 
Should we speak about conformance?

**Comment by Zejnilovic** [comment]
> The latest changes should also fix #491

It does fix the number 2. But it does not fix #491 completely. 
image

**Comment by lokm01** [comment]
> #491

Ok let's deal with that separately.
